dmscripts/publish_draft_services.py

```diff
@@ -0,0 +1,219 @@
+    live_assets_endpoint,
+    client,
+    framework_slug,
+    draft_service,
+    service_id,
+    dry_run=True,
+):
+    if not document_keys:
+        return
+
+    document_updates = {}
+    for document_key in document_keys:
+        if document_key in draft_service:
+            parsed_document_url = _parse_document_url(draft_service[document_key], framework_slug)
+
+            if str(parsed_document_url['supplier_id']) != str(draft_service['supplierId']):
+                raise ValueError(
+                    f"supplier id mismatch: {str(parsed_document_url['supplier_id'])!r} != "
+                    f"{str(draft_service['supplierId'])!r}"
+                )
+            if str(parsed_document_url['draft_id']) != str(draft_service['id']):
+                raise ValueError(
+                    f"draft id mismatch: {str(parsed_document_url['draft_id'])!r} != "
+                    f"{str(draft_service['id'])!r}"
+                )
+
+            draft_document_path = _get_draft_document_path(parsed_document_url, framework_slug)
+            live_document_path = _get_live_document_path(parsed_document_url, framework_slug, service_id)
+
+            try:
+                _copy_document(draft_bucket, documents_bucket, draft_document_path, live_document_path, dry_run)
+
+            except ValueError as e:
+                if not str(e).startswith('Target key already exists in S3'):
+                    raise e
+
+            document_updates[document_key] = urljoin(live_assets_endpoint, live_document_path)
+
+    if dry_run:
+        get_logger().info(
+            "supplier %s: draft %s: dry run: skipped updating document URLs: %s",
+            draft_service["supplierId"],
+            draft_service['id'],
+            document_updates,
+        )
+    else:
+        client.update_service(service_id, document_updates, 'publish_draft_services.py')
+
+
+def _publish_draft_service(
+    client,
+    draft_service,
+    framework_slug,
+    copy_documents_callable=None,
+    dry_run=True,
+    skip_docs_if_published=True,
+):
+    get_logger().info("supplier %s: draft %s: publishing", draft_service["supplierId"], draft_service['id'])
+
+    if draft_service.get("serviceId"):
+        service_id = draft_service["serviceId"]
+        get_logger().warning(
+            "supplier %s: draft %s: skipped publishing - already has service id: %s",
+            draft_service["supplierId"],
+            draft_service['id'],
+            service_id,
+        )
+        if skip_docs_if_published:
+            return
+    elif dry_run:
+        service_id = random.randint(55500000, 55599999)
+        get_logger().info(
+            "supplier %s: draft %s: dry run: generating random test service id: %s",
+            draft_service["supplierId"],
+            draft_service['id'],
+            service_id,
+        )
+    else:
+        try:
+            services = client.publish_draft_service(draft_service['id'], 'publish_draft_services.py')
```

Comment:
Should we make it clearer that `publish_draft_services.py` is the `updated_by` user? I know it's not a named kwarg on the client so maybe a comment, or sticking this in a suitably named variable...

Reply:
Sure.

Reply:
...but... hm... really I feel it should give space to put the *actual user* here which the script should probably take as an argument. But if it's being run from jenkins...? In theory we could try and extract the jenkins user name of the person that hit the button...